### PR TITLE
Lock in current `GitVersionUtils` behaviour with comprehensive tests

### PR DESCRIPTION
**Pull request by FinlayRJW**
## Before this PR

`GitVersionUtils`' tests covered the basic linear cases but left most real-world topology unbounded — RC chains, merges, release branches with backports, multi-tag commits.

#714 discussed [replacing the implementation](https://github.com/palantir/gradle-revapi/pull/714#discussion_r3234807820) with `git for-each-ref --sort=-version:refname`. Before swapping, the suite needs to pin down what current behaviour actually is.

## After this PR

Adds tests across four areas:

- **RC / pre-release chains** — reverse topological order; commit order not semver order; topology not commit-date order; multiple tags on one commit yield one entry (siblings dropped by `describe --abbrev=0`).
- **Merges** — side-branch tag visible after non-ff merge (`describe` walks all parents); tag on a merge commit is returned; squash merge does *not* expose the feature-branch tag.
- **Release branches** — hotfix on `release/1.x` invisible from `develop`; HEAD on `release/1.x` sees pre-fork develop tags but not post-fork; long-lived release branch with RC + hotfixes follows its own chain.
- **HEAD / tag-shape edge cases** — tag at HEAD excluded; annotated tags walk identically to lightweight

No production code changes.

==COMMIT_MSG==
Lock in current `GitVersionUtils` behaviour with tests before the refactor in #714.
==COMMIT_MSG==

## Possible downsides?

A few tests pin arguably strict behaviour e.g. is it true we should always walk in topological order

### CODE REVIEW

File: src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java [hunk 1]
```diff
@@ -111,6 +112,254 @@ void strips_tags_of_v_prefixes(GradleInvoker gradle, Git git) {
...
+        assertOldVersions(gradle, "[1.0.0, 1.0.0-rc1, 1.0.0-rc0]");
+    }
+
+    /*
+     *  o ── o (2.0.0) ── o (1.0.0) ── o (HEAD)
+     *
+     * Tags emerge in commit order; the walker never reshuffles by semver.
+     */
+    @Test
+    void returns_tags_in_topological_order_not_version_string_order(GradleInvoker gradle, Git git) {
```
Comment (FinlayRJW) [inline src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java:137]: fails for `git for-each-ref --sort=-version:refname` approach do we want to keep this behaviour?

File: src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java [hunk 2]
```diff
@@ -111,6 +112,254 @@ void strips_tags_of_v_prefixes(GradleInvoker gradle, Git git) {
         assertOldVersions(gradle, "[1.2.3]");
     }
 
+    /*
+     *  o ── o (1.0.0-rc0) ── o (1.0.0-rc1) ── o (1.0.0) ── o (HEAD)
+     */
+    @Test
+    void returns_rc_chain_in_reverse_topological_order(GradleInvoker gradle, Git git) {
```
Comment (FinlayRJW) [inline src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java:119]: fails for `git for-each-ref --sort=-version:refname` approach do we want to keep this behaviour?

File: src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java [hunk 3]
```diff
@@ -111,6 +112,254 @@ void strips_tags_of_v_prefixes(GradleInvoker gradle, Git git) {
...
+        assertOldVersions(gradle, "[1.0.0, 2.0.0]");
+    }
+
+    /*
+     *   o (committed Jan, tag 2.0.0) ── o (committed Mar, tag 1.0.0) ── o (HEAD)
+     *
+     * Topology beats author/committer date — guards against a stray `--sort` on describe.
+     */
+    @Test
+    void returns_tags_in_topological_order_not_commit_date_order(GradleInvoker gradle, Git git) {
```
Comment (FinlayRJW) [inline src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java:153]: fails for `git for-each-ref --sort=-version:refname` approach do we want to keep this behaviour?

File: src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java [hunk 4]
```diff
@@ -111,6 +112,254 @@ void strips_tags_of_v_prefixes(GradleInvoker gradle, Git git) {
...
+        assertOldVersions(gradle, "[1.0.0, 2.0.0]");
+    }
+
+    /*
+     *  o ── o (1.0.0-rc1 + 1.0.0) ── o (HEAD)
+     *
+     * describe returns one tag per commit, so the sibling is silently dropped.
+     */
+    @Test
+    void multiple_tags_on_same_commit_yield_a_single_entry(GradleInvoker gradle, Git git) {
```
Comment (FinlayRJW) [inline src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java:175]: fails for `git for-each-ref --sort=-version:refname` approach do we want to keep this behaviour?

Comment (FinlayRJW) [reply src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java:175]: when RC + final point at the same commit (e.g. 1.2.1-rc0 + 1.2.1), the current impl silently dropped one because describe --abbrev=0 returns one tag per commit, not sure if this is desired or a bug

File: src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java (hunk 1 above)
Comment (FinlayRJW) [reply src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java:137]: this feels like a bit of an unrealistic test, think that we could just drop this rather than lock in?

File: src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java (hunk 3 above)
Comment (FinlayRJW) [reply src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java:153]: this feels like a bit of an unrealistic test, think that we could just drop this rather than lock in?

File: src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java (hunk 2 above)
Comment (FinlayRJW) [reply src/test/java/com/palantir/gradle/revapi/GitVersionUtilsTest.java:119]: I think this test is both realistic and current behaviour is correct so would need a fix in the `git for-each-ref --sort=-version:refname` world, maybe using `-c  versionsort.suffix=-rc` -- not sure if we would also need to check `-rc, -alpha, -beta, -pre, -snapshot` etc so maybe this is a real limitation?

**Comment by FinlayRJW** [comment]
Merged into https://github.com/palantir/gradle-revapi/pull/714
